File: ЛР3.java
import java.util.*;

// 1. Бинарная куча
class BinaryHeap {
    private List<Integer> heap;
    
    public BinaryHeap() {
        heap = new ArrayList<>();
    }
    
    private int parent(int i) { return (i - 1) / 2; }
    private int leftChild(int i) { return 2 * i + 1; }
    private int rightChild(int i) { return 2 * i + 2; }
    
    public void insert(int key) {
        heap.add(key);
        siftUp(heap.size() - 1);
    }
    
    public Integer extractMin() {
        if (heap.isEmpty()) return null;
        
        int min = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);
        
        if (!heap.isEmpty()) siftDown(0);
        return min;
    }
    
    private void siftUp(int i) {
        while (i > 0 && heap.get(parent(i)) > heap.get(i)) {
            Collections.swap(heap, i, parent(i));
            i = parent(i);
        }
    }
    
    private void siftDown(int i) {
        int minIndex = i;
        int left = leftChild(i);
        int right = rightChild(i);
        
        if (left < heap.size() && heap.get(left) < heap.get(minIndex)) {
            minIndex = left;
        }
        if (right < heap.size() && heap.get(right) < heap.get(minIndex)) {
            minIndex = right;
        }
        
        if (i != minIndex) {
            Collections.swap(heap, i, minIndex);
            siftDown(minIndex);
        }
    }
    
    @Override
    public String toString() {
        return "BinaryHeap: " + heap.toString();
    }
}

// 2. Биномиальная куча (базовая структура)
class BinomialNode {
    int key;
    int degree;
    BinomialNode child;
    BinomialNode sibling;
    
    public BinomialNode(int key) {
        this.key = key;
        this.degree = 0;
    }
}

class BinomialHeap {
    private BinomialNode head;
    
    public BinomialHeap() {
        head = null;
    }
    
    public void insert(int key) {
        BinomialHeap tempHeap = new BinomialHeap();
        tempHeap.head = new BinomialNode(key);
        merge(tempHeap);
    }
    
    public void merge(BinomialHeap other) {
        // Реализация объединения
    }
}

// 3. Хеш-таблица
class HashTable<K, V> {
    private class Entry<K, V> {
        K key;
        V value;
        
        Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }
    
    private int size;
    private LinkedList<Entry<K, V>>[] table;
    
    @SuppressWarnings("unchecked")
    public HashTable(int size) {
        this.size = size;
        table = new LinkedList[size];
        for (int i = 0; i < size; i++) {
            table[i] = new LinkedList<>();
        }
    }
    
    private int hash(K key) {
        return Math.abs(key.hashCode()) % size;
    }
    
    public void put(K key, V value) {
        int index = hash(key);
        for (Entry<K, V> entry : table[index]) {
            if (entry.key.equals(key)) {
                entry.value = value;
                return;
            }
        }
        table[index].add(new Entry<>(key, value));
    }
    
    public V get(K key) {
        int index = hash(key);
        for (Entry<K, V> entry : table[index]) {
            if (entry.key.equals(key)) {
                return entry.value;
            }
        }
        return null;
    }
    
    public void remove(K key) {
        int index = hash(key);
        table[index].removeIf(entry -> entry.key.equals(key));
    }
    
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < size; i++) {
            sb.append(i).append(": ");
            for (Entry<K, V> entry : table[i]) {
                sb.append("(").append(entry.key).append(",").append(entry.value).append(") ");
            }
            sb.append("\n");
        }
        return sb.toString();
    }
}

// Демонстрация
public class HeapDemo {
    public static void main(String[] args) {
        System.out.println("=== Java Demo ===");
        
        // Бинарная куча
        BinaryHeap bh = new BinaryHeap();
        bh.insert(5);
        bh.insert(3);
        bh.insert(8);
        bh.insert(1);
        bh.insert(2);
        System.out.println(bh);
        System.out.println("Extract min: " + bh.extractMin());
        
        // Хеш-таблица
        HashTable<String, Integer> ht = new HashTable<>(10);
        ht.put("apple", 10);
        ht.put("banana", 20);
        ht.put("orange", 30);
        System.out.println("HashTable get apple: " + ht.get("apple"));
        System.out.println(ht);
    }
}

File: ЛР3.cpp
#include <iostream>
#include <vector>
#include <list>
#include <algorithm>
#include <climits>

// 1. Бинарная куча
class BinaryHeap {
private:
    std::vector<int> heap;

    int parent(int i) { return (i - 1) / 2; }
    int leftChild(int i) { return 2 * i + 1; }
    int rightChild(int i) { return 2 * i + 2; }

    void siftUp(int i) {
        while (i > 0 && heap[parent(i)] > heap[i]) {
            std::swap(heap[i], heap[parent(i)]);
            i = parent(i);
        }
    }

    void siftDown(int i) {
        int minIndex = i;
        int left = leftChild(i);
        int right = rightChild(i);

        if (left < heap.size() && heap[left] < heap[minIndex])
            minIndex = left;
        if (right < heap.size() && heap[right] < heap[minIndex])
            minIndex = right;

        if (i != minIndex) {
            std::swap(heap[i], heap[minIndex]);
            siftDown(minIndex);
        }
    }

public:
    void insert(int key) {
        heap.push_back(key);
        siftUp(heap.size() - 1);
    }

    int extractMin() {
        if (heap.empty()) return -1;
        
        int result = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        
        if (!heap.empty()) siftDown(0);
        return result;
    }

    void print() {
        std::cout << "Binary Heap: ";
        for (int val : heap) std::cout << val << " ";
        std::cout << std::endl;
    }
};

// 2. Биномиальная куча (упрощенная реализация)
struct BinomialNode {
    int key;
    int degree;
    BinomialNode* child;
    BinomialNode* sibling;
    
    BinomialNode(int k) : key(k), degree(0), child(nullptr), sibling(nullptr) {}
};

class BinomialHeap {
private:
    BinomialNode* head;

    BinomialNode* mergeTrees(BinomialNode* t1, BinomialNode* t2) {
        if (t1->key > t2->key) std::swap(t1, t2);
        t2->sibling = t1->child;
        t1->child = t2;
        t1->degree++;
        return t1;
    }

public:
    BinomialHeap() : head(nullptr) {}
    
    void insert(int key) {
        BinomialHeap tempHeap;
        tempHeap.head = new BinomialNode(key);
        merge(tempHeap);
    }
    
    void merge(BinomialHeap& other) {
        // Реализация объединения куч
    }
};

// 3. Хеш-таблица с цепочками
class HashTable {
private:
    int size;
    std::vector<std::list<std::pair<int, int>>> table;

    int hash(int key) {
        return key % size;
    }

public:
    HashTable(int s = 10) : size(s), table(s) {}

    void insert(int key, int value) {
        int index = hash(key);
        for (auto& pair : table[index]) {
            if (pair.first == key) {
                pair.second = value;
                return;
            }
        }
        table[index].push_back({key, value});
    }

    int get(int key) {
        int index = hash(key);
        for (auto& pair : table[index]) {
            if (pair.first == key) {
                return pair.second;
            }
        }
        return -1;
    }

    void remove(int key) {
        int index = hash(key);
        table[index].remove_if([key](auto pair) { return pair.first == key; });
    }

    void print() {
        std::cout << "Hash Table:" << std::endl;
        for (int i = 0; i < size; i++) {
            std::cout << i << ": ";
            for (auto& pair : table[i]) {
                std::cout << "(" << pair.first << "," << pair.second << ") ";
            }
            std::cout << std::endl;
        }
    }
};

int main() {
    std::cout << "=== C++ Demo ===" << std::endl;
    
    // Бинарная куча
    BinaryHeap bh;
    bh.insert(5);
    bh.insert(3);
    bh.insert(8);
    bh.insert(1);
    bh.insert(2);
    bh.print();
    std::cout << "Extract min: " << bh.extractMin() << std::endl;
    
    // Хеш-таблица
    HashTable ht;
    ht.insert(1, 100);
    ht.insert(2, 200);
    ht.insert(11, 300); // Коллизия с 1
    ht.print();
    
    return 0;
}
